crates/solver/src/solver/http_solver/instance_cache.rs

```diff
@@ -0,0 +1,187 @@
+use super::{instance_creation::InstanceCreator, settlement::SettlementContext};
+use crate::{s3_instance_upload::S3InstanceUploader, solver::Auction};
+use model::auction::AuctionId;
+use shared::http_solver::model::BatchAuctionModel;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+/// To `Driver` every http solver is presented as an individual `Solver` implementor. Internally
+/// http solvers share the same data that is needed to create the instance for the same auction. In
+/// order to waste less resources we create the instance once per auction in this component.
+pub struct SharedInstanceCreator {
+    creator: InstanceCreator,
+    uploader: Option<Arc<S3InstanceUploader>>,
+    last: Mutex<Option<Cache>>,
+}
+
+struct Cache {
+    run: u64,
+    instances: Instances,
+}
+
+#[derive(Copy, Clone)]
+pub enum InstanceType {
+    Plain,
+    /// without orders that are not connected to the fee token
+    Filtered,
+}
+
+struct Instances {
+    plain: Instance,
+    filtered: Instance,
+}
+
+#[derive(Clone)]
+pub struct Instance {
+    pub model: BatchAuctionModel,
+    pub context: SettlementContext,
+}
+
+impl SharedInstanceCreator {
+    pub fn new(creator: InstanceCreator, uploader: Option<Arc<S3InstanceUploader>>) -> Self {
+        Self {
+            creator,
+            uploader,
+            last: Default::default(),
+        }
+    }
+
+    /// The first call for a new run id creates the instance and stores it in a cache. Subsequent
+    /// calls copy the instance from the cache (and block until the first call completes).
+    pub async fn get_instance(&self, auction: Auction, instance: InstanceType) -> Instance {
```

Comment:
I believe a solver only ever uses a reference to an `Instance`. In that case we could avoid copying those huge structs by storing and returning `Arc<Instance>` here.

Reply:
I like this. Will implement in separate PR.

Reply:
I like the idea. Its a pretty big data structure which takes a lot of memory, `Arc` seems worth it in my eyes.